### ISSUE TITLE
non-fastboot UX -- options for eager loading of known needed assets / default languages?

### ISSUE THREAD

**Issue by NullVoxPopuli** [IgnaceMaes/ember-shiki#17]
tl;dr: I think for the best non-fastboot UX, the primrary languages someone is working with should be eagerly loaded, so there is no async behavior to wait on. This is solved by specifying `defaultLanguages` and `defaultThemes` (if applicable) in the environment.js under the `ember-shiki` ENV config.

This eagerness helps prevents a sort of "flash of unstyled content", which can be jarring to folks looking at the exact place where formatting/highlighting/etc are about to take place.

Also big caveat: I'm working with a _very_ atypical app, a REPL.

------------------------------

I've been exploring ember-shiki and the idea of _lazy load everything_ here: https://github.com/NullVoxPopuli/limber/pull/1303

But I have a problem where the loading experience is noticeably worse than what I had with highlight.js (which was effectively synchronous)

Compare by clicking through a few of the tutorial chapters and paying close attention to how the codeblocks are highlighted
- https://tutorial.glimdown.com/ 
- https://5b230757.limber-glimmer-tutorial.pages.dev/ (uses live [limber.glimdown](https://limber.glimdown.com), rather than also the version in the same PR)

I noticed that there is some config: https://ember-shiki.pages.dev/#configure-(optional)
that _maybe_ sounds like it could solve my problem -- but it's not explicitly stated in the docs? should it be?

```js
    'ember-shiki': {
      defaultLanguages: ['gjs', 'gts', 'css', 'bash', 'js', 'ts', 'markdown'],
      defaultThemes: ['one-dark-pro'],
    },
```

The config _did_ help my rendering UX be less async.

--------------------------------------

_However_, because I'm now using `ember-repl` in the tutorial to generate a component (which is async), there is a bit of a de-sync now between my "scroll-to-top" behavior and rendering of the next page (which, when I used the modifier, highlighting was instant _and_ I could compile *just* markdown synchronously) 😅 -- but this is probably a timing a thing I need to work out anyway. But the compilation stuff is _very heavy_ (babel, template-compiler, etc) -- previously I was only using the unified pipeline with remark and triple-curlies -- so it was very fast :upside_down_face: 

--------------

Some pagespeed insights costs that occurred due to me needing to change how I render prose:

**The Tutorial**

55 on Mobile -- Today (I still have quite a bit to improve on here) https://pagespeed.web.dev/analysis/https-tutorial-glimdown-com/y2voonddn1?form_factor=mobile
43 on Mobile -- the branch where everything is loaded (babel, ember-repl, etc) https://pagespeed.web.dev/analysis/https-fbfe8dfe-limber-glimmer-tutorial-pages-dev/mgcgkfiwcj?form_factor=mobile

**The REPL**
63 on Mobile -- babel is deferred until JS is rendered https://pagespeed.web.dev/analysis/https-limber-glimdown-com/37ehzcf3e4?form_factor=mobile
56 on Mobile -- babel is loaded eagerly https://pagespeed.web.dev/analysis/https-845396c8-limber-glimdown-pages-dev/ksll922uk9?form_factor=mobile

--------------

Additionally, I noticed (before I set the above config) on [the REPL](https://44f1973d.limber-glimdown.pages.dev/edit?c=MQAgMglgtgRgpgJxAUQCYQC4HsEChfCgDqmAFiAFICGAblQMoDGCEADhvo1qnCAGZwAdo14YqAcwDOIRlUEh4IAK6S4qENhDoEcRhgA2ATxA7BPJFyissgoRkkAaXFWkB3OPv0gXJoeY2kvFw8IJKCbKxwGAG8%2BhA0QdyJUHDSfAhYUAB0%2BADy7BA2kgBc%2BAC0IAAGcQmVIBUA%2Bqbm0hDyrPpUIg24FdXxvKw6NBBwrnVNfoit7Z0iDiAdXW3iMaFYSggi3nwYiGs1vM2IPX2Hi8OjrgoeWOMgk2bTIG2Lc3ALS4wrBwO%2BT0h4Po7mtJBstnAerhKjDxAAraSHXAAHj2Vk6ewAfLgQCAABIeYELADicSgKQQAEIUQB6NFLLHQmH4GGVeGIv5DOAjMa4aDWBDRADCmWstkE0XSmRAAHIAALiMkUmmWMV2GUAbj5Vhw0QA3hoEF0ANZqEAAX34GSgsoVSsQdKNjGNK012oF%2BpANgtVul8rgsAdUG4ED4owQbtwcAAHh6tHA%2BFQlPpooxOpJpATPFgiDh9OoY3szNIRTrxfqcSA5Rgnab1FwlBKQABeEAABi1lbazADdhbIAAFABKFuYgIQSRZBtNgDUrYAjJ3caiAwy4NjccvWJiAJobECkWhBOLOs0YQIKJQYbDyPV688TqcbCXmy0YaCpLLImnb-CbkDIjAV43iAd7egARGmEDOuB46Tt2OgpC%2B5qYkKJ7Gt%2BQHXjYG4AXSq4YuuuDmkylS4EAA&format=glimdown) (which is fixed now, due to specifying explicit languages), that if you wait for the preview text to highlight (in the editor area) (before you move the mouse, trackpad, or have any touch/keyboard inputs (at which point the editor would load and replace the preview)), the tab freezes, where it didn't previously :thinking:  
I also noticed that for _Markdown_, it seems shiki loads every language language ever? it takes quite a while!
Is there a way to configure which markdown languages and aliases are loaded (while using async everything?)? maybe this is just something to "be aware of" when not specifying pre-loaded languages?
![image](https://github.com/IgnaceMaes/ember-shiki/assets/199018/277ba5ed-c698-4703-bc46-a672b73dfc72)

**Comment by IgnaceMaes** [comment]
Currently the functionality offered by `CodeBlock` is to encapsulate all async behaviour to ease integration, as you can use the component without configuring anything else and _it just works_.

The API offered by Shiki is async, so if there would be an synchronous way to use `CodeBlock` it would expect that Shiki is already properly initialized and themes/languages are loaded before it's invoked. Initializing Shiki once before rendering anything could be achieved by e.g. doing the initializing in an application initializer, or in a `beforeModel` hook on a route.

This is something you could already do:
```js
async beforeModel() {
  await this.shiki.initialize.perform();
}
```

That would load all predefined languages and themes already. The `CodeBlock` component would still not render 100% synchronously due to the way it currently uses the modifier which performs an async task. Perhaps here there should be a way to make it render instantly assuming the theme/language is loaded.

Overall Shiki would still be less performant compared to Highlight.js in the sense that it uses [vscode-oniguruma](https://github.com/microsoft/vscode-oniguruma) for tokenization which is quite heavy. It's more optimized for correctness and features, rather than file size in browser.

---

Regarding Markdown loading all languages, this is expected behaviour as all those languages are defined as `embeddedLanguage` [in Shiki](https://github.com/shikijs/shiki/blob/main/packages/shiki/src/languages.ts#L848-L856). One way to solve this would be to use a custom grammar definition as mentioned in https://github.com/IgnaceMaes/ember-shiki/issues/18#issuecomment-1704047598.

**Comment by NullVoxPopuli** [comment]
Sounds good!
I just thought also that it'd be a good idea to make sure that padding, margin, and font properties (family, size, line height, kerning, etc) all match the theme that's about to be loaded (likely only if not using a custom theme) -- that'd probably help quite a bit with the noticeable content shiftmonce highlighting happens.

Do you think it'd be good to have a tiny section in the docs about this? Or is it something people should just know, and i forgot? Lol

**Comment by IgnaceMaes** [comment]
I was actually _just_ working on that! Added CSS variables for padding and updated the docs (although it could use further elaboration 😄): https://github.com/IgnaceMaes/ember-shiki/pull/19#issuecomment-1704259979

Not sure what you mean with adjusting the properties bases on the theme? As the themes don't include these settings, only the color styling.

**Comment by NullVoxPopuli** [comment]
ah ok, maybe it's more that I needed to make 
https://github.com/IgnaceMaes/ember-shiki/blob/main/ember-shiki/src/styles.css#L8
match my current font?

anywho, thanks for the additional docs!

**Comment by IgnaceMaes** [comment]
> ah ok, maybe it's more that I needed to make https://github.com/IgnaceMaes/ember-shiki/blob/main/ember-shiki/src/styles.css#L8 match my current font?

Yes, that would be the intended way to style the `CodeBlock`!
